src/frontend/src/optimizer/plan_node/batch_vector_search.rs

```diff
@@ -0,0 +1,177 @@
+// Copyright 2025 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use pretty_xmlish::{Pretty, XmlNode};
+use risingwave_common::catalog::{ColumnDesc, Field, Schema};
+use risingwave_common::types::{DataType, StructType};
+use risingwave_pb::batch_plan::PbVectorIndexNearestNode;
+use risingwave_pb::batch_plan::plan_node::NodeBody;
+use risingwave_pb::common::PbDistanceType;
+
+use crate::OptimizerContextRef;
+use crate::catalog::TableId;
+use crate::expr::{ExprDisplay, ExprImpl, InputRef};
+use crate::optimizer::plan_node::expr_visitable::ExprVisitable;
+use crate::optimizer::plan_node::generic::{GenericPlanNode, GenericPlanRef};
+use crate::optimizer::plan_node::utils::{Distill, childless_record};
+use crate::optimizer::plan_node::{
+    Batch, BatchPlanRef as PlanRef, BatchPlanRef, ExprRewritable, PlanBase, PlanTreeNodeUnary,
+    ToBatchPb, ToDistributedBatch, ToLocalBatch,
+};
+use crate::optimizer::property::{Distribution, FunctionalDependencySet, Order};
+
+#[derive(Debug, Clone, educe::Educe)]
+#[educe(Hash, PartialEq, Eq)]
+pub struct BatchVectorSearchCore {
+    pub input: BatchPlanRef,
+    pub top_n: u64,
+    pub distance_type: PbDistanceType,
+    pub index_name: String,
+    pub index_table_id: TableId,
+    pub info_column_desc: Vec<ColumnDesc>,
+    pub vector_column_idx: usize,
+    #[educe(Hash(ignore), Eq(ignore))]
+    pub ctx: OptimizerContextRef,
+}
+
+impl GenericPlanNode for BatchVectorSearchCore {
+    fn functional_dependency(&self) -> FunctionalDependencySet {
+        // FunctionalDependencySet::new(self.info_column_desc.len())
+        // TODO: include dependency derived from info columns
+        self.input.functional_dependency().clone()
+    }
+
+    fn schema(&self) -> Schema {
+        let mut schema = self.input.schema().clone();
+        schema.fields.push(Field::new(
+            "vector_info",
+            DataType::List(
+                DataType::Struct(StructType::new(
+                    self.info_column_desc
+                        .iter()
+                        .map(|col| (col.name.clone(), col.data_type.clone()))
+                        .chain([("__distance".to_owned(), DataType::Float64)]),
+                ))
+                .into(),
+            ),
+        ));
+        schema
```

Comment:
Is unnesting the vector search result a common next step of processing? If yes, I think we can flatten the schema to allow one input row to generate multiple rows.